File: frontend/src/types/platform-fee.ts
export interface FeeRateConfig {
  id: string;
  transaction_type: string;
  fee_type: string;
  rate: string;
  min_fee: string;
  max_fee: string | null;
  effective_from: string | null;
  effective_until: string | null;
  created_at: string;
  updated_at: string;
}

export interface FeeRateConfigCreate {
  transaction_type: string;
  fee_type?: string;
  rate?: string;
  min_fee?: string;
  max_fee?: string;
  effective_from?: string;
  effective_until?: string;
}

export interface FeeRateConfigUpdate {
  rate?: string;
  min_fee?: string;
  max_fee?: string;
  effective_from?: string;
  effective_until?: string;
}

export interface PlatformFee {
  id: string;
  network_transaction_id: string;
  fee_rate_config_id: string | null;
  fee_type: string;
  rate: string;
  base_amount: string;
  calculated_amount: string;
  currency: string;
  status: "pending" | "collected" | "waived";
  collected_at: string | null;
  waived_by: string | null;
  waived_reason: string | null;
  created_at: string;
}

export interface PaginatedFees {
  items: PlatformFee[];
  total: number;
  page: number;
  per_page: number;
}

export interface FeeStats {
  total_fees: number;
  pending_amount: string;
  collected_amount: string;
  waived_amount: string;
  total_revenue: string;
}

File: backend/app/schemas/platform_fee.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Fee Rate Config schemas
# ---------------------------------------------------------------------------


class FeeRateConfigCreate(BaseModel):
    transaction_type: str = Field(
        ..., description="order, invoice, payment, case_transfer"
    )
    fee_type: str = Field(
        default="transaction_percent",
        description="transaction_percent, flat_fee, subscription_addon",
    )
    rate: Decimal = Field(default=Decimal("0.0000"))
    min_fee: Decimal = Field(default=Decimal("0.00"))
    max_fee: Decimal | None = None
    effective_from: datetime | None = None
    effective_until: datetime | None = None


class FeeRateConfigUpdate(BaseModel):
    rate: Decimal | None = None
    min_fee: Decimal | None = None
    max_fee: Decimal | None = None
    effective_from: datetime | None = None
    effective_until: datetime | None = None


class FeeRateConfigResponse(BaseModel):
    id: str
    transaction_type: str
    fee_type: str
    rate: Decimal
    min_fee: Decimal
    max_fee: Decimal | None = None
    effective_from: datetime | None = None
    effective_until: datetime | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Platform Fee schemas
# ---------------------------------------------------------------------------


class PlatformFeeResponse(BaseModel):
    id: str
    network_transaction_id: str
    fee_rate_config_id: str | None = None
    fee_type: str
    rate: Decimal
    base_amount: Decimal
    calculated_amount: Decimal
    currency: str
    status: str
    collected_at: datetime | None = None
    waived_by: str | None = None
    waived_reason: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class PaginatedFees(BaseModel):
    items: list[PlatformFeeResponse]
    total: int
    page: int
    per_page: int


class WaiveFeeRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=255)


class FeeStats(BaseModel):
    total_fees: int
    pending_amount: Decimal
    collected_amount: Decimal
    waived_amount: Decimal
    total_revenue: Decimal
